examples/pipeline/wav2letter.py

```diff
@@ -0,0 +1,1179 @@
+
+    base_name = os.path.basename(filename)
+
+    # We want to ignore anything after '_nohash_' in the file name when
+    # deciding which set to put a wav in, so the data set creator has a way of
+    # grouping wavs that are close variations of each other.
+    hash_name = re.sub(r'_nohash_.*$', '', base_name).encode("utf-8")
+
+    # This looks a bit magical, but we need to decide whether this file should
+    # go into the training, testing, or validation sets, and we want to keep
+    # existing files in the same set even if more files are subsequently
+    # added.
+    # To do that, we need a stable way of deciding based on just the file name
+    # itself, so we do a hash of that and then use that to generate a
+    # probability value that we use to assign it.
+    hash_name_hashed = hashlib.sha1(hash_name).hexdigest()
+    percentage_hash = ((int(hash_name_hashed, 16) % (
+        MAX_NUM_WAVS_PER_CLASS + 1)) * (100.0 / MAX_NUM_WAVS_PER_CLASS))
+
+    if percentage_hash < validation_percentage:
+        result = 'validation'
+    elif percentage_hash < (testing_percentage + validation_percentage):
+        result = 'testing'
+    else:
+        result = 'training'
+
+    return result
+
+
+def filter_speechcommands(tag, training_percentage, data):
+    if training_percentage < 100.:
+        testing_percentage = (
+            100. - training_percentage - validation_percentage)
+
+        def which_set_filter(x):
+            return which_set(x, validation_percentage, testing_percentage) == tag
+
+        data._walker = list(filter(which_set_filter, data._walker))
+    return data
+
+
+def datasets_speechcommands():
+
+    root = "./"
+
+    def create(tag):
+        data = SPEECHCOMMANDS(root, download=True)
+        data = filter_speechcommands(tag, training_percentage, data)
+        data = Processed(process_datapoint, data)
+        # data = diskcache_iterator(data)
+        data = MapMemoryCache(data)
+        return data
+
+    return create("training"), create("validation"), create("testing")
+
+
+if args.dataset == "librispeech":
+    training, validation, _ = datasets_librispeech()
+elif args.dataset == "speechcommand":
+    training, validation, _ = datasets_speechcommands()
+
+
+# Word Decoder
+
+
+def greedy_decode(outputs):
+    """Greedy Decoder. Returns highest probability of class labels for each timestep
+
+    Args:
+        outputs (torch.Tensor): shape (input length, batch size, number of classes (including blank))
+
+    Returns:
+        torch.Tensor: class labels per time step.
+    """
+    _, indices = topk(outputs, k=1, dim=-1)
+    return indices[..., 0]
+
+
+def build_transitions():
+
+    from collections import Counter
+
+    c = None
+
+    for _, label in training:
+        # Count bigrams
+        count = [((a.item(), b.item())) for (a, b) in zip(label, label[1:])]
+        count = Counter(count)
+        if c is None:
+            c = count
+        else:
+            c = c + count
```

Comment:
I think you could also use update and initialize c as an empty Counter.

Creating the Counter and adding to it over each iteration might be quite slow in comparison.